File: 언어별_정리/python/배열/two_pointers_sliding_window.py
"""
투 포인터 / 슬라이딩 윈도우 예제.
- 배열이 모두 양수일 때 길이/합 조건을 빠르게 맞출 수 있다.
"""

from typing import List


def longest_subarray_leq(arr: List[int], limit: int) -> int:
    """
    합이 limit 이하인 가장 긴 구간 길이.
    양수 배열 기준 O(N) 슬라이딩 윈도우.
    """

    left = 0
    cur = 0
    best = 0
    for right, val in enumerate(arr):
        cur += val
        while cur > limit:
            cur -= arr[left]
            left += 1
        best = max(best, right - left + 1)
    return best


if __name__ == "__main__":
    nums = [2, 1, 3, 2, 4]
    print(longest_subarray_leq(nums, 6))  # 3 (2+1+3)

File: 언어별_정리/java/배열/TwoPointersSlidingWindow.java
import java.util.*;

public class TwoPointersSlidingWindow {
    public static void main(String[] args) {
        // 설명: 투 포인터/슬라이딩 윈도우로 합이 S 이상인 최소 길이
        // 시간복잡도: O(N)
        // 주의사항: 모든 값이 양수일 때만 슬라이딩 윈도우가 유효
        // 예시 입력(하드코딩):
        // nums = [2, 3, 1, 2, 4, 3], S = 7
        // 예상 출력:
        // minLen=2 (부분합 4+3)

        int[] nums = {2, 3, 1, 2, 4, 3};
        int s = 7;

        int n = nums.length;
        int left = 0;
        int sum = 0;
        int best = Integer.MAX_VALUE;

        for (int right = 0; right < n; right++) {
            sum += nums[right];
            while (sum >= s) {
                best = Math.min(best, right - left + 1);
                sum -= nums[left++];
            }
        }

        int ans = (best == Integer.MAX_VALUE) ? 0 : best;
        System.out.println("minLen=" + ans);
    }
}
